client/DESIGN.md

```diff
@@ -0,0 +1,380 @@
+
+### view 
+
+The view maintains the map of operations performed throughout the Transaction.  This map has type table:key -> (op, value).  When the user commits the transactions, each updated value in the view is sent to the cache, which writes the changes through to the DB server via batch write.  (If the view's values are out-of-date with the DB, write will throw a StaleException.)
+
+Note that every value read from the cache during the Transaction is stored in this map.  Even values that do not change within the Transaction are stored in the map with the *hold* operation, as we explain below.
+
+Unlike the Cache, the Transaction does not maintain a history of values.  Rather, it stores the user's operations on each entry.
+
+Consider an arbitrary entry in the DB of the form 
+    table:key -> (op, value)  
+The possible ops in the map are the following:
+
+#### create
+Add a new entry to the DB with key *key* and value *value*.
+
+#### hold
+Maintain the current value in the DB.  The *hold* operation does not require a value, i.e., `table:key -> (hold, None)`, because we are not changing the value.
+
+The hold operation exists because the computed output of our Transaction may be dependent on any value that we read from Cache during the Transaction.  Therefore, the success of the Transaction depends on these values having NOT changed in the DB between the time we read and commit them.  If these value have changed, the Transaction is stale and should throw a StaleException.  The *hold* operation allows us to maintain this invariant.
+
+#### update
+Replace the given key's old value with the new value provided.
+
+#### delete
+Remove the given key from the database.  The *delete* operation does not require a value, i.e., `table:key -> (delete, None)`, because we are not changing the value.
+
+## Transaction Public Methods
+
+Here is the hierarchy of cache-lookup on a user request.  Each of the methods below (1) first looks to the view for the values it needs to complete its operation.  (2) If the values are not in the view, it attempts to contact the cache to complete its operation.  (3) If the cache does not contain the necessary information, the cache makes the HTTP request to the HTTP server.  (The Transaction does not directly make HTTP requests.)
+
+### read
+    read(
+        table: string, 
+        key: string, 
+        max_age=None, 
+        no_cache=False) returns value (JSON) or None
+
+The read method first attempts to read the specified key from the view, then from cache, and finally from the DB server.  If read finds the requested value, it stores it in the view; otherwise, it returns None and stores the key had no value (in Python, None) for that read_timestamp satisfying the given parameters.
+
+Note that all Transaction reads have the same read timestamp, which is specified in the constructor (unlike the Cache read method).
```

Comment:
The read method can throw a StaleException when this read gets a value whose value_time > read_time of some value read earlier.